knowledge_repo/app/index.py

```diff
@@ -19,8 +24,58 @@
 UNLOCKED = '0'
 
 
+def set_up_indexing_timers(app):
+    if not app.config['INDEXING_ENABLED']:
+        return False
+
+    if app.check_thread_support(check_index=True, check_repositories=app.config['INDEXING_UPDATES_REPOSITORIES']):
+        if os.environ['KNOWLEDGE_REPO_MASTER_UUID'] != app.uuid:
+            logger.info("Not spawning index-sync timers for non-master application instance: {}".format(app.uuid))
+            return
+
+        def index_watchdog(app):
+            while True:
+                if not hasattr(app, 'sync_thread') or not app.sync_thread.is_alive():
+                    logger.warning("Master indexing thread has died. Restarting...")
+                    context = None
+                    with app.app_context():
+                        app.sync_thread = multiprocessing.Process(target=index_sync_loop, args=(app,))
+                        app.sync_thread.start()
+                time.sleep(app.config['INDEXING_TIMEOUT'])
+
+        def index_sync_loop(app):
+            while True:
+                with app.app_context():
+                    update_index(check_timeouts=False)
+                time.sleep(app.config['INDEXING_INTERVAL'])
+
+        app.index_watchdog = multiprocessing.Process(target=index_watchdog, args=(app,))
+        app.index_watchdog.start()
+    else:
+        @app.before_request
+        def update_index_if_required():
+            update_index()
+
+
+def acquire_index_lock():
+    if IndexMetadata.get('lock', 'master_check', None) is None:
+        try:
+            IndexMetadata.set('lock', 'master_check', current_app.uuid)
+            db_session.commit()
+            if (
+                IndexMetadata.get('lock', 'index_master') == current_app.uuid or
+                time_since(IndexMetadata.get_last_update('lock', 'index_master'), default=current_app.config["INDEXING_TIMEOUT"] + 1) > current_app.config["INDEXING_TIMEOUT"]
+            ):
+                IndexMetadata.set('lock', 'index_master', current_app.uuid)  # Update/set lock and update timestamp
+                return True
+        finally:
+            IndexMetadata.set('lock', 'master_check', None)
```

Comment:
I'm a little confused by the master_check. regardless of what happens in the `try:` statement, don't we immediately set the master_check back to none?

Reply:
That's correct... but it means that only one thread can attempt to become the master indexer at any time, preventing race conditions that may allow two processes to both be the master indexer for the first subsequent sync operation. It's unlikely to happen, but coding defensively.